src/uct/cuda/cuda_ipc/cuda_ipc_iface.c

```diff
@@ -343,22 +338,33 @@ static ucs_mpool_ops_t uct_cuda_ipc_event_desc_mpool_ops = {
     .obj_cleanup   = uct_cuda_ipc_event_desc_cleanup,
 };
 
+ucs_status_t uct_cuda_ipc_unmap_memhandle(void *mapped_addr)
+{
+    return UCT_CUDADRV_FUNC(cuIpcCloseMemHandle((CUdeviceptr)mapped_addr));
+}
+
 ucs_status_t uct_cuda_ipc_map_memhandle(void *arg, uct_cuda_ipc_key_t *key,
                                         void **mapped_addr)
 {
+    if (key->unmap_memhandle_func == NULL) {
+        key->unmap_memhandle_func = uct_cuda_ipc_unmap_memhandle;
+    }
+
```

Comment:
It seems like the assumption here is that there's a 1:1 mapping between rkey and cuda IPC memhandle. For concurrent operations that use the same memhandle but touch different offsets, isn't the assumption invalidated? If so, you could still end up trying to close the memhandle when there's still an outstanding operation. Correct me if I'm mistaken. Maintaining a recount for each unique memhandle is probably safer.

Reply:
As far as I could understand, there is a 1:1 mapping between the rkey and the memhandle. If you touch different offsets (which is actually what happens when transferring blocks > 1GB), the memhandle will not be closed until the immediately before the rkey is release, that is, before the `ucs_free((void *)rkey)` call.

Reply:
Consider the following test:
```
char *ptr;
cudaMalloc(&ptr, 1024 * 1024 * 1024 * sizeof(char)); // yields memhandle1
if (rank == 1) {
    ucp_tag_recv_nb(worker, ptr, 4096, ...) // rkey1 points to memhandle1
    ucp_tag_recv_nb(worker, ptr + 4096, 4096, ...) // rkey2 also points to memhandle1
} else /* rank 0 */ {
    ucp_tag_send_nb(ep_to_rank1, ptr, 4096, ...) // consumes rkey1 in worker_progress, sets local_key1->unmap
    ucp_tag_send_nb(ep_to_rank1, ptr + 4096, 4096, ...) // consumes rkey2 in worker_progress, sets local_key2->unmap
    ucp_worker_progress() 
   /* when 1st send completes, local_key1->unmap maybe called
    * and closememhandle can be called when 2nd send may still be
    * in progress */
}
```
@pentschev As I understand it, because there are separate instances of key structures for the essentially the same remote memhandle at rank 0, the PR code doesn't prevent calling closememhandle on a memhandle that may still be active.

Reply:
Sorry @Akshay-Venkatesh , I completely missed your comment here. The case fixed here is when we have a single rkey mapping a single memhandle, but where copies are triggered in various smaller chunks (rather than a single copy operation that copies the buffer at once). In such case, `uct_cuda_ipc_progress_event_q` would perform the copies of each chunk and unmap the handle for each one of them, which would crash when the copy involves multiple events in the queue (the first event already unmapped the memhandle).

TBH, I didn't even know having multiple rkeys mapping the same memhandle was allowed. I haven't seen that case in my tests so far, but I can see how that would be a problem and we would then need a refcount. Assuming that having multiple rkeys mapping the same memhandle is a normal use case (I'm not very familiar with that), then I think we should solve it in a future PR.